modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,197 @@
+        for {
+          snapshots <- getSnapshotsCountPerEpoch(timeTriggerInterval)
+          epoch <- getEpochNumber(snapshotOrdinal, snapshots)
+          reward <- getTotalRewardsPerEpoch(epoch)
+          // Note: We need rounding there to match validator rewards sheet rounding
+          rewardPerSnapshot <- NonNegLong
+            .from(Math.round(reward.value / snapshots.toDouble))
+            .left
+            .map(RewardsException)
+        } yield Amount(rewardPerSnapshot)
+
+      private[rewards] def getSnapshotsCountPerEpoch(
+        timeTriggerInterval: FiniteDuration
+      ): Either[RewardsError, PosLong] =
+        PosLong
+          .from((config.epochDurationInYears * 12 * 30 * 24 * 60).toLong / timeTriggerInterval.toMinutes)
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getTotalRewardsPerEpoch(epochNumber: PosInt): Either[RewardsError, Amount] =
+        NonNegLong
+          .from(
+            (config.baseEpochReward.value / Math.pow(2, epochNumber.toDouble - 1.0)).toLong
+          )
+          .map(Amount(_))
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getEpochNumber(
+        snapshotOrdinal: SnapshotOrdinal,
+        snapshotsPerEpoch: PosLong
+      ): Either[RewardsError, PosInt] =
+        snapshotOrdinal match {
+          case ordinal if ordinal == SnapshotOrdinal.MinValue => Right(1)
+          case ordinal =>
+            for (epoch <- 1 to config.epochs)
+              if (ordinal.value > (epoch - 1) * snapshotsPerEpoch && ordinal.value <= epoch * snapshotsPerEpoch)
+                return PosInt.from(epoch).left.map(RewardsException)
+            Left(LastEpochExceeded)
+
+        }
+
+      private[rewards] def createSharedDistribution(
+        snapshotHash: Hash,
+        snapshotReward: Amount,
+        addresses: NonEmptySet[Address]
+      ): F[Either[RewardsError, NonEmptyMap[Address, Amount]]] = {
+        val totalSpace = NonNegInt.unsafeFrom(addresses.length)
+        val (quotient, remainder) = snapshotReward.coerce.toLong /% totalSpace.toLong
+
+        val quotientDistribution = addresses.toNonEmptyList.map(_ -> quotient).toList.toMap.pure[F]
+
+        val remainderDistribution = if (remainder > 0) {
+          Random.scalaUtilRandomSeedInt(snapshotHash.value.toInt).flatMap { random =>
+            random.shuffleList(addresses.toNonEmptyList.toList).map { shuffled =>
+              shuffled.take(remainder.toInt).map(_ -> remainder / remainder).toMap
+            }
+          }
+        } else Map.empty[Address, Long].pure[F]
+
+        (quotientDistribution, remainderDistribution)
+          .mapN(_ |+| _)
+          .map(_.view.mapValues(NonNegLong.unsafeFrom).mapValues(Amount(_)).toList)
+          .map(map => NonEmptyMap.fromMapUnsafe(SortedMap.from(map)))
+          .map(_.asRight[RewardsError])
+          .handleError(error => RewardsException(error.getMessage).asLeft[NonEmptyMap[Address, Amount]])
```

Comment:
Lot's of unsafe and throwing operations, where there shouldn't be many. I think you can rewrite it to:

```
val quotientDistribution = addresses.toNonEmptyList.map(_ -> NonNegLong.unsafeFrom(quotient)).toNem

Random.scalaUtilRandomSeedInt(snapshotHash.value.hashCode).flatMap { random =>
  random
    .shuffleList(addresses.toList)
    .map { shuffled =>
      shuffled.take(remainder.toInt)
    }
    .map { top =>
      top.toNel
        .fold(quotientDistribution) { topNel =>
          topNel.map(_ -> NonNegLong(1L)).toNem |+| quotientDistribution
        }
        .mapBoth { case (addr, reward) => (addr, Amount(reward)) }
    }
}
```